storage/database/db_manager_test.go

```diff
@@ -161,28 +168,177 @@ func TestDBManager_FastTrieProgress(t *testing.T) {
+	encodedBody, err := rlp.EncodeToBytes(body)
+	if err != nil {
+		t.Fatal("Failed to encode body!", "err", err)
+	}
+
+	for _, dbm := range dbManagers {
+		assert.False(t, dbm.HasBody(hash1, num1))
+		assert.Nil(t, dbm.ReadBody(hash1, num1))
+		assert.Nil(t, dbm.ReadBodyInCache(hash1))
+		assert.Nil(t, dbm.ReadBodyRLP(hash1, num1))
+		assert.Nil(t, dbm.ReadBodyRLPByHash(hash1))
+
+		dbm.WriteBody(hash1, num1, body)
+
+		assert.True(t, dbm.HasBody(hash1, num1))
+		assert.Equal(t, body, dbm.ReadBody(hash1, num1))
+		assert.Equal(t, body, dbm.ReadBodyInCache(hash1))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLP(hash1, num1))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLPByHash(hash1))
+
+		dbm.DeleteBody(hash1, num1)
+
+		assert.False(t, dbm.HasBody(hash1, num1))
+		assert.Nil(t, dbm.ReadBody(hash1, num1))
+		assert.Nil(t, dbm.ReadBodyInCache(hash1))
+		assert.Nil(t, dbm.ReadBodyRLP(hash1, num1))
+		assert.Nil(t, dbm.ReadBodyRLPByHash(hash1))
+
+		dbm.WriteBodyRLP(hash1, num1, encodedBody)
+
+		assert.True(t, dbm.HasBody(hash1, num1))
+		assert.Equal(t, body, dbm.ReadBody(hash1, num1))
+		assert.Equal(t, body, dbm.ReadBodyInCache(hash1))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLP(hash1, num1))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLPByHash(hash1))
+
+		bodyBatch := dbm.NewBatch(BodyDB)
+		dbm.PutBodyToBatch(bodyBatch, hash2, num2, body)
+		if err := bodyBatch.Write(); err != nil {
+			t.Fatal("Failed to write batch!", "err", err)
+		}
+
+		assert.True(t, dbm.HasBody(hash2, num2))
+		assert.Equal(t, body, dbm.ReadBody(hash2, num2))
+		assert.Equal(t, body, dbm.ReadBodyInCache(hash2))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLP(hash2, num2))
+		assert.Equal(t, rlp.RawValue(encodedBody), dbm.ReadBodyRLPByHash(hash2))
+	}
 }
 
+// TestDBManager_Td tests read, write and delete operations of blockchain headers' total difficulty.
 func TestDBManager_Td(t *testing.T) {
-	// TODO-Klaytn-Database Implement this!
+	for _, dbm := range dbManagers {
+		assert.Nil(t, dbm.ReadTd(hash1, num1))
+
+		dbm.WriteTd(hash1, num1, big.NewInt(12345))
+		assert.Equal(t, big.NewInt(12345), dbm.ReadTd(hash1, num1))
+
+		dbm.WriteTd(hash1, num1, big.NewInt(54321))
+		assert.Equal(t, big.NewInt(54321), dbm.ReadTd(hash1, num1))
+
+		dbm.DeleteTd(hash1, num1)
+		assert.Nil(t, dbm.ReadTd(hash1, num1))
+	}
 }
 
+// TestDBManager_Receipts read, write and delete operations of blockchain receipts.
 func TestDBManager_Receipts(t *testing.T) {
-	// TODO-Klaytn-Database Implement this!
+	header := &types.Header{Number: big.NewInt(int64(num1))}
+	headerHash := header.Hash()
+	receipts := types.Receipts{generateReceipt(111)}
+
+	for _, dbm := range dbManagers {
+		assert.Nil(t, dbm.ReadReceipts(headerHash, num1))
+		assert.Nil(t, dbm.ReadReceiptsByBlockHash(headerHash))
+
+		dbm.WriteReceipts(headerHash, num1, receipts)
+		dbm.WriteHeader(header)
```

Comment:
What does this line do? When reading Receipt, is there a task to read header from DB for receipt?